mysys_ssl/yassl.cc

```diff
@@ -0,0 +1,194 @@
+/*
+ Copyright (c) 2015 MariaDB Corporation
+
+ This program is free software; you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation; version 2 of the License.
+
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program; if not, write to the Free Software
+ Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA */
+
+/*
+  The very minimal subset of OpenSSL's EVP* functions.
+  Just enough for my_crypt.cc to work.
+
+  On the other hand, where it has to implement OpenSSL functionality,
+  it tries to be compatible (e.g. same flags and struct member names).
+*/
+
+#include <openssl/ssl.h>
+#include "aes.hpp"
+
+using yaSSL::yaERR_remove_state;
+
+#define EVP_CIPH_ECB_MODE     0x1
+#define EVP_CIPH_CBC_MODE     0x2
+#define EVP_CIPH_NO_PADDING 0x100
+
+/*
+  note that TaoCrypt::AES object is not explicitly put into EVP_CIPHER_CTX.
+  That's because we need to control when TaoCrypt::AES constructor and
+  destructor are called.
+*/
+typedef struct
```

Comment:
Are values encrypted with YaSSL binary compatible with those generated by OpenSSL. Especially taking into account our custom scheme for encrypting data of size which is not a multiple of AES_BLOCK_SIZE.
In other words, will we be able to read binary log generated on a server with YaSSL on a server with OpenSSL? And vice versa.

Reply:
There would be a quite devastating problem for yassl if they have implemented AES differently than openssl did. AES is a standard. Just to be double sure I have conducted an experiment :): 
Run encrypted_master.test on openssl build. Started PS build with yassl, flush logs, copied over binlog files. PS was able to see only first mysql-bin.00001, but that was not a problem as I copied all binlog files as mysql-bin.00001 one by one. Then used mysqlbinlog with --read-from-remote-server - it was able to decrypt binlog files. Long story short both PS with yassl can decrypt encrypted binlog produced by PS with openssl and vice versa.